### PR TITLE
Add support for http proxy

### PR DESCRIPTION
**Pull request by moderakh**
Adds support for Http Proxy when reading data from storage

sample configuration:

```
spark.delta.sharing.network.proxyHost=1.2.3.4
spark.delta.sharing.network.proxyPort=3128
spark.delta.sharing.network.noProxyHosts=5.6.7.8,12.13.14.15
```

### CODE REVIEW

File: client/src/main/scala/io/delta/sharing/client/DeltaSharingFileSystem.scala [hunk 1]
```diff
@@ -22,37 +22,78 @@ import java.util.concurrent.TimeUnit
...
       .disableAutomaticRetries()
-      .build()
+
+    // Set proxy if provided
+    proxyConfigOpt.foreach { proxyConfig =>
+      val proxy = new HttpHost(proxyConfig.host, proxyConfig.port)
+      clientBuilder.setProxy(proxy)
+
+      // Set credentials if provided
+      if (proxyConfig.username.isDefined && proxyConfig.password.isDefined) {
```
Comment (chakankardb) [inline client/src/main/scala/io/delta/sharing/client/DeltaSharingFileSystem.scala:70]: What if only one is defined?

File: client/src/main/scala/io/delta/sharing/client/DeltaSharingFileSystem.scala [hunk 2]
```diff
@@ -22,37 +22,78 @@ import java.util.concurrent.TimeUnit
...
+              // Route via proxy
+              new HttpRoute(target, null, proxy, true)
+            }
+          }
+        }
+        clientBuilder.setRoutePlanner(routePlanner)
+      }
+    }
+    clientBuilder.build()
   }
```
Comment (chakankardb) [inline client/src/main/scala/io/delta/sharing/client/DeltaSharingFileSystem.scala:97]: Is there a way to unit test this?

Comment (moderakh) [reply client/src/main/scala/io/delta/sharing/client/DeltaSharingFileSystem.scala:97]: I will add unit test, I was waiting to agree on the general approach before adding tests.

File: client/src/main/scala/io/delta/sharing/client/DeltaSharingFileSystem.scala (hunk 1 above)
Comment (moderakh) [reply client/src/main/scala/io/delta/sharing/client/DeltaSharingFileSystem.scala]: removed username/pass as not needed.

File: client/src/main/scala/io/delta/sharing/client/DeltaSharingFileSystem.scala (hunk 2 above)
Comment (moderakh) [reply client/src/main/scala/io/delta/sharing/client/DeltaSharingFileSystem.scala:88]: added unit test with a locally hosted http server

File: client/src/main/scala/io/delta/sharing/client/DeltaSharingFileSystem.scala [hunk 3]
```diff
@@ -22,37 +22,69 @@ import java.util.concurrent.TimeUnit
...
+    val clientBuilder = HttpClientBuilder.create()
       .setMaxConnTotal(maxConnections)
       .setMaxConnPerRoute(maxConnections)
       .setDefaultRequestConfig(config)
       // Disable the default retry behavior because we have our own retry logic.
       // See `RetryUtils.runWithExponentialBackoff`.
       .disableAutomaticRetries()
-      .build()
+
+    // Set proxy if provided
```
Comment (chakankardb) [inline client/src/main/scala/io/delta/sharing/client/DeltaSharingFileSystem.scala:65]: nit: periods at the end of comments.

File: client/src/test/scala/io/delta/sharing/client/util/ConfUtilsSuite.scala [hunk 4]
```diff
@@ -90,4 +90,35 @@ class ConfUtilsSuite extends SparkFunSuite {
...
+    assert(proxyConfig.get.host == "1.2.3.4")
+    assert(proxyConfig.get.port == 8080)
+    assert(proxyConfig.get.noProxyHosts.isEmpty)
+  }
+
+  test("getProxyConfig with no proxy settings") {
+    val conf = newConf()
+    val proxyConfig = getProxyConfig(conf)
+    assert(proxyConfig.isEmpty)
+  }
```
Comment (chakankardb) [inline client/src/test/scala/io/delta/sharing/client/util/ConfUtilsSuite.scala:123]: Consider adding negative tests:  Invalid port, null/empty host.

File: client/src/test/scala/io/delta/sharing/client/DeltaSharingFileSystemSuite.scala [hunk 5]
```diff
@@ -58,4 +65,56 @@ class DeltaSharingFileSystemSuite extends SparkFunSuite {
...
+      // Step 5: Configure the httpClient to use the ProxyConfig
+      // Please replace this with actual code to configure the httpClient
+
+      // Step 6: Send a request to the local server through the httpClient
+      val response = httpClient.execute(new HttpGet(server.getURI.toString))
+
+      // Step 7: Assert that the request is successful
+      assert(response.getStatusLine.getStatusCode == HttpServletResponse.SC_OK)
+      val content = EntityUtils.toString(response.getEntity)
+      assert(content.trim == "Hello, World!")
```
Comment (chakankardb) [inline client/src/test/scala/io/delta/sharing/client/DeltaSharingFileSystemSuite.scala:114]: Can we test NO_PROXY_HOST ?

File: client/src/test/scala/io/delta/sharing/client/util/ConfUtilsSuite.scala (hunk 4 above)
Comment (moderakh) [reply client/src/test/scala/io/delta/sharing/client/util/ConfUtilsSuite.scala:123]: added negative tests

File: client/src/test/scala/io/delta/sharing/client/DeltaSharingFileSystemSuite.scala (hunk 5 above)
Comment (moderakh) [reply client/src/test/scala/io/delta/sharing/client/DeltaSharingFileSystemSuite.scala:115]: added tests for noProxyHosts

File: client/src/main/scala/io/delta/sharing/client/DeltaSharingFileSystem.scala (hunk 3 above)
Comment (moderakh) [reply client/src/main/scala/io/delta/sharing/client/DeltaSharingFileSystem.scala]: addressed

File: client/src/test/scala/io/delta/sharing/client/DeltaSharingFileSystemSuite.scala [hunk 6]
```diff
@@ -58,4 +64,117 @@ class DeltaSharingFileSystemSuite extends SparkFunSuite {
...
+      // Send a request to the local server through the httpClient.
+      val response = httpClient.execute(new HttpGet(server.getURI.toString))
+
+      // Assert that the request is successful.
+      assert(response.getStatusLine.getStatusCode == HttpServletResponse.SC_OK)
+      val content = EntityUtils.toString(response.getEntity)
+      assert(content.trim == "Hello, World!")
+
+      // Assert that the request is not passed through proxy.
+      assert(proxyServer.getCapturedRequests().isEmpty)
```
Comment (chakankardb) [inline client/src/test/scala/io/delta/sharing/client/DeltaSharingFileSystemSuite.scala:174]: nice!

**Review by chakankardb — APPROVED** [review]
LGTM!

**Review by linzhou-db — COMMENTED** [review]
overall looks good though there are some lint nits.
